### PR TITLE
Subscribe to navigation events on open files

### PR DESCRIPTION
**Pull request by JoeyScarr**
Use that to navigate if available, otherwise fall back to requesting navigation directly from the analysis server.

### CODE REVIEW

**Comment by JoeyScarr** [comment]
This is the change I mentioned in #3620.

File: src/extension/providers/dart_reference_provider.ts [hunk 1]
```diff
@@ -33,11 +34,16 @@ export class DartReferenceProvider implements ReferenceProvider, DefinitionProvi
 	}
 
 	public async provideDefinition(document: TextDocument, position: Position, token: CancellationToken): Promise<DefinitionLink[] | undefined> {
-		const resp = await this.analyzer.analysisGetNavigation({
-			file: fsPath(document.uri),
-			length: 0,
-			offset: document.offsetAt(position),
-		});
+		let resp1 = this.fileTracker.getNavigationTargets(fsPath(document.uri), document.offsetAt(position));
```
Comment (DanTup) [inline src/extension/providers/dart_reference_provider.ts:37]: nit: Is there any reason to name this `resp1` and not just `resp` and remove the assignment on line 46?

File: src/extension/providers/dart_reference_provider.ts [hunk 2]
```diff
@@ -33,11 +34,16 @@ export class DartReferenceProvider implements ReferenceProvider, DefinitionProvi
...
+		if (!resp1) {
+			resp1 = await this.analyzer.analysisGetNavigation({
+				file: fsPath(document.uri),
+				length: 0,
+				offset: document.offsetAt(position),
+			});
+		}
+		if (!resp1) return undefined;
```
Comment (DanTup) [inline src/extension/providers/dart_reference_provider.ts:45]: Is it possible for this to be undefined/null at this point? The original code seems to assume `analysisGetNavigation` can't return undefined, so I think it should still be the case after the above has run?

**Review by DanTup — COMMENTED** [review]
I think this looks fine (with two minor nits/comments), although I think it may introduce a race that could be hit if the server is busy that might result in inconsistent results:

- Notification is sent from server, client caches it
- Server gets busy so it's not responding to requests
- Modify a file (notification is sent to server)
- Try to navigate - it will use the stale results from the previous notification (since the server hasn't sent the new one yet)

With the previous code, the request would go to the server and be queued/async, so should return the latest results.

I'm not sure this is an issue worth worrying about though (both because it seems unlikely, and because almost all users are using LSP), so I don't think it blocks landing this.

I'll wait for your response to the comments below, but otherwise I think we can merge this. Thanks for the contribution!

File: src/extension/providers/dart_reference_provider.ts (hunk 1 above)
Comment (JoeyScarr) [reply src/extension/providers/dart_reference_provider.ts:37]: Probably not. I just did it to satisfy the type system; if you remove this it seems to want non-null assertions on the usages of `resp` below. That's probably better.

File: src/extension/providers/dart_reference_provider.ts (hunk 2 above)
Comment (JoeyScarr) [reply src/extension/providers/dart_reference_provider.ts:45]: Same as above, was just to satisfy the type system. Feel free to remove, you'll just have to add a couple of `1`s in the code below.

**Comment by JoeyScarr** [comment]
> I think this looks fine (with two minor nits/comments), although I think it may introduce a race that could be hit if the server is busy that might result in inconsistent results:
> 
> * Notification is sent from server, client caches it
> * Server gets busy so it's not responding to requests
> * Modify a file (notification is sent to server)
> * Try to navigate - it will use the stale results from the previous notification (since the server hasn't sent the new one yet)
> 
> With the previous code, the request would go to the server and be queued/async, so should return the latest results.
> 
> I'm not sure this is an issue worth worrying about though (both because it seems unlikely, and because almost all users are using LSP), so I don't think it blocks landing this.
> 
> I'll wait for your response to the comments below, but otherwise I think we can merge this. Thanks for the contribution!

There is another issue I think, which is that if the DAS is slow to send back the navigations for a file after it's modified, the stored navigation offsets will be wrong. This could be solved by tracking offset changes from edits and taking them into account when finding the right navigation link to use, but will leave up to you to decide whether you care.

**Comment by DanTup** [comment]
> There is another issue I think, which is that if the DAS is slow to send back the navigations for a file after it's modified, the stored navigation offsets will be wrong. This could be solved by tracking offset changes from edits and taking them into account when finding the right navigation link to use,

Yeah, that's the same issue really. Another option could be to just blow away the cached data whenever a file is modified, so in that case it'll hit the server. It could be done with a listener on `workspace.onDidChangeTextDocument` in the constructor, similar to the existing open/close calls.

I'm happy to make that change after merging this, or you're welcome to add (or, if you think that will introduce problems, let me know).

File: src/extension/providers/dart_reference_provider.ts (hunk 2 above)
Comment (DanTup) [reply src/extension/providers/dart_reference_provider.ts:45]: I'd prefer not to have the `!` so this is fine - I just thought TypeScript was a little cleverer and wouldn't have reported it 🙄

**Comment by JoeyScarr** [comment]
> > There is another issue I think, which is that if the DAS is slow to send back the navigations for a file after it's modified, the stored navigation offsets will be wrong. This could be solved by tracking offset changes from edits and taking them into account when finding the right navigation link to use,
> 
> Yeah, that's the same issue really. Another option could be to just blow away the cached data whenever a file is modified, so in that case it'll hit the server. It could be done with a listener on `workspace.onDidChangeTextDocument` in the constructor, similar to the existing open/close calls.
> 
> I'm happy to make that change after merging this, or you're welcome to add (or, if you think that will introduce problems, let me know).

Sounds fine, feel free to add that. I've been using the plugin with just this patch and the issue doesn't seem to crop up too often, so I'm not too fussed either way. Thanks for the review!

**Comment by DanTup** [comment]
I decided not to bother with the change above, as I don't think it would be 100% reliable anyway. Eg:

- you modify document (1)
- we clear the local cache
- you modify document (2)
- we clear the local cache
- server sends notification from edit (1)
- you Go-to-Definition, and use the stale results

So since we can't really guard against it reliably, I'm not sure it's worth trying to cover a subset of those cases. The chance of it occurring seems fairly slim anyway (and ofc this code will ultimately go away).

Thanks for the contribution!

**Pull request merged**